core/mutex.c

```diff
@@ -60,19 +59,68 @@ int mutex_lock(mutex_t *mutex)
         thread_add_to_list(&mutex->queue, me);
     }
 
-
     irq_restore(irq_state);
     thread_yield_higher();
     /* We were woken up by scheduler. Waker removed us from queue. */
-    return 0;
+}
+
+void mutex_lock(mutex_t *mutex)
+{
+    unsigned irq_state = irq_disable();
+
+    DEBUG("PID[%" PRIkernel_pid "]: mutex_lock()\n", thread_getpid());
+
+    if (mutex->queue.next == NULL) {
+        /* mutex is unlocked. */
+        mutex->queue.next = MUTEX_LOCKED;
+        DEBUG("PID[%" PRIkernel_pid "]: mutex_lock() early out.\n",
+              thread_getpid());
+        irq_restore(irq_state);
+    }
+    else {
+        _block(mutex, irq_state);
+    }
+}
+
+int mutex_lock_cancelable(mutex_cancel_t *mc)
+{
+    unsigned irq_state = irq_disable();
+
+    DEBUG("PID[%" PRIkernel_pid "]: mutex_lock_cancelable()\n",
+          thread_getpid());
+
+    if (mc->cancelled) {
```

Comment:
```suggestion
    if (mc->cancelled) {
        mc->cancelled = 0;
```

so we can re-use the mutex and cancel it again

Reply:
This won't work. The usual use case looks like this:

```C
static void timeout_cb(void *_arg) {
    mutex_cancel(arg);
}

int ztimer_mutex_lock_timeout(ztimer_clock_t *clock, mutex_t *mutex,
                              uint32_t timeout)
{
    mutex_cancel_t mc = mutex_cancel_init(mutex);
    ztimer_t t;
    t.callback = timeout_cb;
    t.arg = &mc;
    ztimer_set(clock, &t, timeout);
    if (0 == mutex_lock_cancelable(mutex)) {
        /* BANG! */
        ztimer_remove(clock, &t);
        return 0;
    }
    return -ECANCELED;
}
```

Let's say the timer expired at `/* BANG! */`. In that case `mc->cancel` will be set to `1` and `mutex_lock_cancelable()` has no chance to reset it to zero. The `mutex_cancel_t` structure must always be reset only after the IRQ triggering the cancel is disabled (or the thread sending the cancel is notified to no longer cancel the mutex).

Reply:
Ok that makes sense. So `mutex_cancel_init()` marks the earliest time a mutex can be canceled (also for subsequent calls)